### DISCUSSION TITLE
useContractWrite never prepares (stays undefined) and other issues

### DISCUSSION THREAD

**Discussion by theljwhite** [wevm/wagmi#2871] [Q&A]
Hey, sorry if this is the wrong place to ask this. I'm having a couple issues while attempting to write from contract. Admittedly, it is my first time performing this specific action with wagmi. I have an implementation for browser connected users to mint (write) with ethers.js. But what I am trying to do here is give more options to mint with Web3Modal/wallet connect.

I've noticed that on initial render, my write function seemingly never prepares and stays undefined which hints at a potential timing or initialization issue. There is a lot of async stuff (where I am checking contract state with ethers first before my mint function is reached) but I have reviewed my code and cant find anything that I'm doing wrong. 

Note that if I force a re-render (edit the file and save it works flawlessly) the write function prepares and I am able to proceed with the entire process and successfully mint. So far I have only tested with Web3Modal > WalletConnect > Metamask connection on mobile view in Google Chrome on Polygon Mumbai...and it is a nextJS app.  

I am using usePrepareContractWrite combined with useContractWrite. However, if I bypass the prepare hook and attempt to only use useContractWrite, the write function is available on initial render but I get this error when the write function is called:

 `ConnectorNotFoundError: Connector not found
    at prepareWriteContract (chunk-LAFZBYO7.js:2063:1)
    at async writeContract (chunk-LAFZBYO7.js:2354:1)`  

But once again, if I use the combination of the two, I get no errors, but I experience the issue I described where the write function stays undefined.

Here is my setup for the hooks. If need be, I can provide more of my code as well: 

In the body of my Mint.tsx component: 

```
const { config: wcWhitelistMintConfig } = usePrepareContractWrite({
    address: myContractAddress,
    abi: contractAbi,
    functionName: "whitelistMint",
    args: [[searchQuery], txReferrer],
    value: isProduction
      ? ethers.utils.parseUnits("1", "ether").toBigInt()
      : ethers.utils.parseUnits("100", "wei").toBigInt(),
  });

  const {
    write: walletConnectWhitelistMint,
    data: wcWhitelistMintData,
    error: wcWhitelistMintError,
  } = useContractWrite(wcWhitelistMintConfig);
```
The write function is called within another function that fires when a user clicks a button
(but of course on initial render I cant click the button as I have disabled={!walletConnectWhitelistMint}) haha 

But the function that runs onClick is basically just this: 
`if (activeSale === "whitelist") walletConnectWhitelistMint?.();`

I can provide more of the code or try to expand on other details if need be. I've made sure that the searchQuery and txReferrer args I am passing in are not undefined before this component is even rendered and I tried passing in a simplified version of my contractABI just to see but the issues remain. Although I dont think any params I am passing in are an issue, as it works perfectly once the component has been forced to re-render. Thanks!

**Comment by theljwhite** [comment]
Didnt want to open this up as an issue in case I'm overlooking something or it's a non wagmi related issue, so any opinions appreciated haha.

**Comment by theljwhite** [comment]
After working on other stuff for a few days, I came back to this and fixed it. Posting this in case someone else has a similar issue. I added refetch to my prepare hook: 

`
const { config: wcWhitelistMintConfig, refetch: refecthWhitelistWcMint } =
    usePrepareContractWrite({
      ...mintHookParams,
      functionName: "whitelistMint",
      onError(error) {
        handleWalletConnectMintErrors(error);
      },
    });`

and by simply calling the refetch function in a useEffect with empty dependency array, everything is working and the write function from useContractWrite is immediately usable via button click.

`useEffect(() => {
    refetchWalletConnectMint();
  }, []);`

**Reply by ogubuikeAlex** [reply]
I have this same Issue.
But in my case I am calling write inside a useEffect hook and adding refetch doesn't solve it.
